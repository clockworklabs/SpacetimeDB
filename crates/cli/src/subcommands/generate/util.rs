//! Various utility functions that the generate modules have in common.

use std::fmt::{Display, Formatter, Result};

/// Turns a closure `f: Fn(&mut Formatter) -> Result` into `fmt::Display`.
pub(super) fn fmt_fn(f: impl Fn(&mut Formatter) -> Result) -> impl Display {
    struct FDisplay<F>(F);
    impl<F: Fn(&mut Formatter) -> Result> Display for FDisplay<F> {
        fn fmt(&self, f: &mut Formatter<'_>) -> Result {
            (self.0)(f)
        }
    }
    FDisplay(f)
}
