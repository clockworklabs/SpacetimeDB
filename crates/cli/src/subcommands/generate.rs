#![warn(clippy::uninlined_format_args)]

use anyhow::Context;
use clap::parser::ValueSource;
use clap::Arg;
use clap::ArgAction::Set;
use fs_err as fs;
use spacetimedb_codegen::{generate, Csharp, Go, Lang, Rust, TypeScript, AUTO_GENERATED_PREFIX};
use spacetimedb_lib::de::serde::DeserializeWrapper;
use spacetimedb_lib::{sats, RawModuleDef};
use spacetimedb_schema;
use spacetimedb_schema::def::ModuleDef;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use crate::tasks::csharp::dotnet_format;
use crate::tasks::go::go_format;
use crate::tasks::rust::rustfmt;
use crate::util::{resolve_sibling_binary, y_or_n};
use crate::Config;
use crate::{build, common_args};
use clap::builder::PossibleValue;
use std::collections::BTreeSet;
use std::io::Read;

pub fn cli() -> clap::Command {
    clap::Command::new("generate")
        .about("Generate client files for a spacetime module.")
        .override_usage("spacetime generate --lang <LANG> --out-dir <DIR> [--project-path <DIR> | --bin-path <PATH>]")
        .arg(
            Arg::new("wasm_file")
                .value_parser(clap::value_parser!(PathBuf))
                .long("bin-path")
                .short('b')
                .group("source")
                .conflicts_with("project_path")
                .conflicts_with("build_options")
                .help("The system path (absolute or relative) to the compiled wasm binary we should inspect"),
        )
        .arg(
            Arg::new("project_path")
                .value_parser(clap::value_parser!(PathBuf))
                .default_value(".")
                .long("project-path")
                .short('p')
                .group("source")
                .help("The system path (absolute or relative) to the project you would like to inspect"),
        )
        .arg(
            Arg::new("json_module")
                .hide(true)
                .num_args(0..=1)
                .value_parser(clap::value_parser!(PathBuf))
                .long("module-def")
                .group("source")
                .help("Generate from a ModuleDef encoded as json"),
        )
        .arg(
            Arg::new("out_dir")
                .value_parser(clap::value_parser!(PathBuf))
                .required(true)
                .long("out-dir")
                .short('o')
                .help("The system path (absolute or relative) to the generate output directory"),
        )
        .arg(
            Arg::new("namespace")
                .default_value("SpacetimeDB.Types")
                .long("namespace")
                .help("The namespace that should be used"),
        )
        .arg(
            Arg::new("lang")
                .required(true)
                .long("lang")
                .short('l')
                .value_parser(clap::value_parser!(Language))
                .help("The language to generate"),
        )
        .arg(
            Arg::new("build_options")
                .long("build-options")
                .alias("build-opts")
                .action(Set)
                .default_value("")
                .help("Options to pass to the build command, for example --build-options='--lint-dir='"),
        )
        .arg(common_args::yes())
        .after_help("Run `spacetime help publish` for more detailed information.")
}

pub async fn exec(config: Config, args: &clap::ArgMatches) -> anyhow::Result<()> {
    exec_ex(config, args, extract_descriptions).await
}

/// Like `exec`, but lets you specify a custom a function to extract a schema from a file.
pub async fn exec_ex(
    config: Config,
    args: &clap::ArgMatches,
    extract_descriptions: ExtractDescriptions,
) -> anyhow::Result<()> {
    let project_path = args.get_one::<PathBuf>("project_path").unwrap();
    let wasm_file = args.get_one::<PathBuf>("wasm_file").cloned();
    let json_module = args.get_many::<PathBuf>("json_module");
    let out_dir = args.get_one::<PathBuf>("out_dir").unwrap();
    let lang = *args.get_one::<Language>("lang").unwrap();
    let namespace = args.get_one::<String>("namespace").unwrap();
    let force = args.get_flag("force");
    let build_options = args.get_one::<String>("build_options").unwrap();

    if args.value_source("namespace") == Some(ValueSource::CommandLine) && lang != Language::Csharp {
        return Err(anyhow::anyhow!("--namespace is only supported with --lang csharp"));
    }

    let module: ModuleDef = if let Some(mut json_module) = json_module {
        let DeserializeWrapper::<RawModuleDef>(module) = if let Some(path) = json_module.next() {
            serde_json::from_slice(&fs::read(path)?)?
        } else {
            serde_json::from_reader(std::io::stdin().lock())?
        };
        module.try_into()?
    } else {
        let wasm_path = if let Some(path) = wasm_file {
            println!("Skipping build. Instead we are inspecting {}", path.display());
            path.clone()
        } else {
            build::exec_with_argstring(config.clone(), project_path, build_options).await?
        };
        let spinner = indicatif::ProgressBar::new_spinner();
        spinner.enable_steady_tick(std::time::Duration::from_millis(60));
        spinner.set_message("Extracting schema from wasm...");
        extract_descriptions(&wasm_path).context("could not extract schema")?
    };

    fs::create_dir_all(out_dir)?;

    let mut paths = BTreeSet::new();

    let csharp_lang;
    let go_lang;
    let gen_lang = match lang {
        Language::Csharp => {
            csharp_lang = Csharp { namespace };
            &csharp_lang as &dyn Lang
        }
        Language::Go => {
            go_lang = Go::default();
            &go_lang as &dyn Lang
        }
        Language::Rust => &Rust,
        Language::TypeScript => &TypeScript,
    };

    for (fname, code) in generate(&module, gen_lang) {
        let fname = Path::new(&fname);
        // If a generator asks for a file in a subdirectory, create the subdirectory first.
        if let Some(parent) = fname.parent().filter(|p| !p.as_os_str().is_empty()) {
            fs::create_dir_all(out_dir.join(parent))?;
        }
        let path = out_dir.join(fname);
        fs::write(&path, code)?;
        paths.insert(path);
    }

    // TODO: We should probably just delete all generated files before we generate any, rather than selectively deleting some afterward.
    let mut auto_generated_buf: [u8; AUTO_GENERATED_PREFIX.len()] = [0; AUTO_GENERATED_PREFIX.len()];
    let files_to_delete = walkdir::WalkDir::new(out_dir)
        .into_iter()
        .map(|entry_result| {
            let entry = entry_result?;
            // Only delete files.
            if !entry.file_type().is_file() {
                return Ok(None);
            }
            let path = entry.into_path();
            // Don't delete regenerated files.
            if paths.contains(&path) {
                return Ok(None);
            }
            // Only delete files that start with the auto-generated prefix.
            let mut file = fs::File::open(&path)?;
            Ok(match file.read_exact(&mut auto_generated_buf) {
                Ok(()) => (auto_generated_buf == AUTO_GENERATED_PREFIX.as_bytes()).then_some(path),
                Err(err) if err.kind() == std::io::ErrorKind::UnexpectedEof => None,
                Err(err) => return Err(err.into()),
            })
        })
        .filter_map(Result::transpose)
        .collect::<anyhow::Result<Vec<_>>>()?;

    if !files_to_delete.is_empty() {
        println!("The following files were not generated by this command and will be deleted:");
        for path in &files_to_delete {
            println!("  {}", path.to_str().unwrap());
        }

        if y_or_n(force, "Are you sure you want to delete these files?")? {
            for path in files_to_delete {
                fs::remove_file(path)?;
            }
            println!("Files deleted successfully.");
        } else {
            println!("Files not deleted.");
        }
    }

    if let Err(err) = lang.format_files(paths) {
        // If we couldn't format the files, print a warning but don't fail the entire
        // task as the output should still be usable, just less pretty.
        eprintln!("Could not format generated files: {err}");
    }

    println!("Generate finished successfully.");
    Ok(())
}

#[derive(Clone, Copy, PartialEq)]
pub enum Language {
    Csharp,
    Go,
    TypeScript,
    Rust,
}

impl clap::ValueEnum for Language {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::Csharp, Self::Go, Self::TypeScript, Self::Rust]
    }
    fn to_possible_value(&self) -> Option<PossibleValue> {
        Some(match self {
            Self::Csharp => clap::builder::PossibleValue::new("csharp").aliases(["c#", "cs"]),
            Self::Go => clap::builder::PossibleValue::new("go").aliases(["golang"]),
            Self::TypeScript => clap::builder::PossibleValue::new("typescript").aliases(["ts", "TS"]),
            Self::Rust => clap::builder::PossibleValue::new("rust").aliases(["rs", "RS"]),
        })
    }
}

impl Language {
    fn format_files(&self, generated_files: BTreeSet<PathBuf>) -> anyhow::Result<()> {
        match self {
            Language::Rust => rustfmt(generated_files)?,
            Language::Csharp => dotnet_format(generated_files)?,
            Language::Go => go_format(generated_files)?,
            Language::TypeScript => {
                // TODO: implement formatting.
            }
        }

        Ok(())
    }
}

pub type ExtractDescriptions = fn(&Path) -> anyhow::Result<ModuleDef>;
fn extract_descriptions(wasm_file: &Path) -> anyhow::Result<ModuleDef> {
    let bin_path = resolve_sibling_binary("spacetimedb-standalone")?;
    let child = Command::new(&bin_path)
        .arg("extract-schema")
        .arg(wasm_file)
        .stdout(Stdio::piped())
        .spawn()
        .with_context(|| format!("failed to spawn {}", bin_path.display()))?;
    let sats::serde::SerdeWrapper::<RawModuleDef>(module) = serde_json::from_reader(child.stdout.unwrap())?;
    Ok(module.try_into()?)
}
