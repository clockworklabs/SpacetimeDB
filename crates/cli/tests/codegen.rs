use spacetimedb_cli::generate;
use spacetimedb_data_structures::map::HashMap;
use spacetimedb_testing::modules::{CompilationMode, CompiledModule};
use std::path::Path;
use std::sync::OnceLock;

fn compiled_module() -> &'static Path {
    static COMPILED_MODULE: OnceLock<CompiledModule> = OnceLock::new();
    COMPILED_MODULE
        .get_or_init(|| CompiledModule::compile("rust-wasm-test", CompilationMode::Debug))
        .path()
}

macro_rules! declare_tests {
    ($($name:ident => $lang:ident,)*) => {
        $(
            #[test]
            fn $name() {
                let module = generate::extract_descriptions(compiled_module()).unwrap();
                let outfiles: HashMap<_, _> = generate::generate(module, generate::Language::$lang, "SpacetimeDB")
                    .unwrap()
                    .into_iter()
                    .collect();
                insta::with_settings!({ sort_maps => true }, {
                    insta::assert_toml_snapshot!(outfiles);
                });
            }
        )*
    }
}

declare_tests! {
    test_codegen_csharp => Csharp,
    test_codegen_typescript => TypeScript,
    test_codegen_rust => Rust,
}
