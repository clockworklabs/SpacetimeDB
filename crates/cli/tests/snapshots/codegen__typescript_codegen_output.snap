---
source: crates/cli/tests/codegen.rs
expression: outfiles
---
"add_player_reducer.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export class AddPlayerReducer
{
	public static call(name: string)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.call("add_player", [name]);
		}
	}

	public static deserializeArgs(rawArgs: any[]): any[] {
		let nameType = AlgebraicType.createPrimitiveType(BuiltinType.Type.String);
		let nameValue = AlgebraicValue.deserialize(nameType, rawArgs[0])
		let name = nameValue.asString();
		return [name];
	}

	public static on(callback: (status: string, identity: string, reducerArgs: any[]) => void)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.on("reducer:AddPlayer", callback);
		}
	}
}

__SPACETIMEDB__.reducers.set("AddPlayer", AddPlayerReducer);
if (__SPACETIMEDB__.spacetimeDBClient) {
	__SPACETIMEDB__.spacetimeDBClient.registerReducer("AddPlayer", AddPlayerReducer);
}

export default AddPlayerReducer
'''
"namespace_test_c.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, SumTypeVariant, BuiltinType, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export namespace Namespace.testC {
	export function getAlgebraicType(): AlgebraicType {
		return AlgebraicType.createSumType([
			new SumTypeVariant("Foo", AlgebraicType.createProductType([
		])),
			new SumTypeVariant("Bar", AlgebraicType.createProductType([
		])),
		]);
	}

	export function serialize(value: Namespace.testC): object {
		const result: {[key: string]: any} = {};
		result[value.tag] = [];
		return result;
	}

	export type Foo = { tag: "Foo"; value: undefined };
	export type Bar = { tag: "Bar"; value: undefined };

	export function fromValue(value: AlgebraicValue): Namespace.testC {
		let sumValue = value.asSumValue();
		let tag = sumValue.tag;
		let variant = Namespace.testC.getAlgebraicType().sum.variants[tag];
		return { tag: variant.name, value: undefined } as Namespace.testC;
	}
}

export type Namespace.testC = Namespace.testC.Foo | Namespace.testC.Bar;
export default Namespace.testC;
'''
"repeating_test_reducer.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export class RepeatingTestReducer
{
	public static call(prevTime: number)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.call("repeating_test", [prevTime]);
		}
	}

	public static deserializeArgs(rawArgs: any[]): any[] {
		let prevTimeType = AlgebraicType.createPrimitiveType(BuiltinType.Type.U64);
		let prevTimeValue = AlgebraicValue.deserialize(prevTimeType, rawArgs[0])
		let prevTime = prevTimeValue.asNumber();
		return [prevTime];
	}

	public static on(callback: (status: string, identity: string, reducerArgs: any[]) => void)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.on("reducer:RepeatingTest", callback);
		}
	}
}

__SPACETIMEDB__.reducers.set("RepeatingTest", RepeatingTestReducer);
if (__SPACETIMEDB__.spacetimeDBClient) {
	__SPACETIMEDB__.spacetimeDBClient.registerReducer("RepeatingTest", RepeatingTestReducer);
}

export default RepeatingTestReducer
'''
"test_a.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, SumType, SumTypeVariant, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export class TestA extends IDatabaseTable
{
	public static tableName = "TestA";
	public x: number;
	public y: number;
	public z: string;

	constructor(x: number, y: number, z: string) {
	super();
		this.x = x;
		this.y = y;
		this.z = z;
	}

	public static serialize(value: TestA): object {
		return [
		value.x, value.y, value.z
		];
	}

	public static getAlgebraicType(): AlgebraicType
	{
		return AlgebraicType.createProductType([
			new ProductTypeElement("x", AlgebraicType.createPrimitiveType(BuiltinType.Type.U32)),
			new ProductTypeElement("y", AlgebraicType.createPrimitiveType(BuiltinType.Type.U32)),
			new ProductTypeElement("z", AlgebraicType.createPrimitiveType(BuiltinType.Type.String)),
		]);
	}

	public static fromValue(value: AlgebraicValue): TestA
	{
		let productValue = value.asProductValue();
		let __x = productValue.elements[0].asNumber();
		let __y = productValue.elements[1].asNumber();
		let __z = productValue.elements[2].asString();
		return new this(__x, __y, __z);
	}

	public static count(): number
	{
		return __SPACETIMEDB__.clientDB.getTable("TestA").count();
	}

	public static all(): TestA[]
	{
		return __SPACETIMEDB__.clientDB.getTable("TestA").getInstances() as unknown as TestA[];
	}

	public static filterByX(value: number): TestA[] | null
	{
		let result: TestA[] = [];
		for(let entry of __SPACETIMEDB__.clientDB.getTable("TestA").getEntries())
		{
			var productValue = entry.asProductValue();
			let compareValue = productValue.elements[0].asNumber() as number;
			if (compareValue == value) {
				result.push(TestA.fromValue(entry));
			}
		}
		return result;
	}

	public static filterByY(value: number): TestA[] | null
	{
		let result: TestA[] = [];
		for(let entry of __SPACETIMEDB__.clientDB.getTable("TestA").getEntries())
		{
			var productValue = entry.asProductValue();
			let compareValue = productValue.elements[1].asNumber() as number;
			if (compareValue == value) {
				result.push(TestA.fromValue(entry));
			}
		}
		return result;
	}

	public static filterByZ(value: string): TestA[] | null
	{
		let result: TestA[] = [];
		for(let entry of __SPACETIMEDB__.clientDB.getTable("TestA").getEntries())
		{
			var productValue = entry.asProductValue();
			let compareValue = productValue.elements[2].asString() as string;
			if (compareValue == value) {
				result.push(TestA.fromValue(entry));
			}
		}
		return result;
	}


	public static onInsert(callback: (value: TestA) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestA").onInsert(callback);
	}

	public static onUpdate(callback: (oldValue: TestA, newValue: TestA) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestA").onUpdate(callback);
	}

	public static onDelete(callback: (value: TestA) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestA").onDelete(callback);
	}

	public static removeOnInsert(callback: (value: TestA) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestA").removeOnInsert(callback);
	}

	public static removeOnUpdate(callback: (oldValue: TestA, newValue: TestA) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestA").removeOnUpdate(callback);
	}

	public static removeOnDelete(callback: (value: TestA) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestA").removeOnDelete(callback);
	}

}

export default TestA;

__SPACETIMEDB__.registerComponent("TestA", TestA);
'''
"test_b.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, SumType, SumTypeVariant, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export class TestB extends IDatabaseTable
{
	public static tableName = "TestB";
	public foo: string;

	public static primaryKey: string | undefined = undefined;

	constructor(foo: string) {
	super();
		this.foo = foo;
	}

	public static serialize(value: TestB): object {
		return [
		value.foo
		];
	}

	public static getAlgebraicType(): AlgebraicType
	{
		return AlgebraicType.createProductType([
			new ProductTypeElement("foo", AlgebraicType.createPrimitiveType(BuiltinType.Type.String)),
		]);
	}

	public static fromValue(value: AlgebraicValue): TestB
	{
		let productValue = value.asProductValue();
		let __foo = productValue.elements[0].asString();
		return new this(__foo);
	}

}

export default TestB;

__SPACETIMEDB__.registerComponent("TestB", TestB);
'''
"test_d.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, SumType, SumTypeVariant, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";
// @ts-ignore
import { Namespace.TestC } from "./namespace.test_c";

export class TestD extends IDatabaseTable
{
	public static tableName = "TestD";
	public testC: Namespace.TestC | null;

	constructor(testC: Namespace.TestC | null) {
	super();
		this.testC = testC;
	}

	public static serialize(value: TestD): object {
		return [
		value.testC ? { "some": Namespace.TestC.serialize(value.testC) } : { "none": [] }
		];
	}

	public static getAlgebraicType(): AlgebraicType
	{
		return AlgebraicType.createProductType([
			new ProductTypeElement("test_c", AlgebraicType.createSumType([
			new SumTypeVariant("some", Namespace.TestC.getAlgebraicType()),
			new SumTypeVariant("none", AlgebraicType.createProductType([
		])),
		])),
		]);
	}

	public static fromValue(value: AlgebraicValue): TestD
	{
		let productValue = value.asProductValue();
		let __test_c = function() { const value = productValue.elements[0].asSumValue().tag == 1 ? null : productValue.elements[0].asSumValue().value; return value ? Namespace.TestC.fromValue(value) : null; }();
		return new this(__test_c);
	}

	public static count(): number
	{
		return __SPACETIMEDB__.clientDB.getTable("TestD").count();
	}

	public static all(): TestD[]
	{
		return __SPACETIMEDB__.clientDB.getTable("TestD").getInstances() as unknown as TestD[];
	}


	public static onInsert(callback: (value: TestD) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestD").onInsert(callback);
	}

	public static onUpdate(callback: (oldValue: TestD, newValue: TestD) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestD").onUpdate(callback);
	}

	public static onDelete(callback: (value: TestD) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestD").onDelete(callback);
	}

	public static removeOnInsert(callback: (value: TestD) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestD").removeOnInsert(callback);
	}

	public static removeOnUpdate(callback: (oldValue: TestD, newValue: TestD) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestD").removeOnUpdate(callback);
	}

	public static removeOnDelete(callback: (value: TestD) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestD").removeOnDelete(callback);
	}

}

export default TestD;

__SPACETIMEDB__.registerComponent("TestD", TestD);
'''
"test_e.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, SumType, SumTypeVariant, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export class TestE extends IDatabaseTable
{
	public static tableName = "TestE";
	public id: number;
	public name: string;

	public static primaryKey: string | undefined = "id";

	constructor(id: number, name: string) {
	super();
		this.id = id;
		this.name = name;
	}

	public static serialize(value: TestE): object {
		return [
		value.id, value.name
		];
	}

	public static getAlgebraicType(): AlgebraicType
	{
		return AlgebraicType.createProductType([
			new ProductTypeElement("id", AlgebraicType.createPrimitiveType(BuiltinType.Type.U64)),
			new ProductTypeElement("name", AlgebraicType.createPrimitiveType(BuiltinType.Type.String)),
		]);
	}

	public static fromValue(value: AlgebraicValue): TestE
	{
		let productValue = value.asProductValue();
		let __id = productValue.elements[0].asNumber();
		let __name = productValue.elements[1].asString();
		return new this(__id, __name);
	}

	public static count(): number
	{
		return __SPACETIMEDB__.clientDB.getTable("TestE").count();
	}

	public static all(): TestE[]
	{
		return __SPACETIMEDB__.clientDB.getTable("TestE").getInstances() as unknown as TestE[];
	}

	public static filterById(value: number): TestE | null
	{
		for(let entry of __SPACETIMEDB__.clientDB.getTable("TestE").getEntries())
		{
			var productValue = entry.asProductValue();
			let compareValue = productValue.elements[0].asNumber() as number;
			if (compareValue == value) {
				return TestE.fromValue(entry);
			}
		}
		return null;
	}

	public static filterByName(value: string): TestE[] | null
	{
		let result: TestE[] = [];
		for(let entry of __SPACETIMEDB__.clientDB.getTable("TestE").getEntries())
		{
			var productValue = entry.asProductValue();
			let compareValue = productValue.elements[1].asString() as string;
			if (compareValue == value) {
				result.push(TestE.fromValue(entry));
			}
		}
		return result;
	}


	public static onInsert(callback: (value: TestE) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestE").onInsert(callback);
	}

	public static onUpdate(callback: (oldValue: TestE, newValue: TestE) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestE").onUpdate(callback);
	}

	public static onDelete(callback: (value: TestE) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestE").onDelete(callback);
	}

	public static removeOnInsert(callback: (value: TestE) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestE").removeOnInsert(callback);
	}

	public static removeOnUpdate(callback: (oldValue: TestE, newValue: TestE) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestE").removeOnUpdate(callback);
	}

	public static removeOnDelete(callback: (value: TestE) => void)
	{
		__SPACETIMEDB__.clientDB.getTable("TestE").removeOnDelete(callback);
	}

}

export default TestE;

__SPACETIMEDB__.registerComponent("TestE", TestE);
'''
"test_reducer.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";
// @ts-ignore
import { TestA } from "./test_a";
// @ts-ignore
import { TestB } from "./test_b";
// @ts-ignore
import { Namespace.TestC } from "./namespace.test_c";

export class TestReducer
{
	public static call(arg: TestA, arg2: TestB, arg3: Namespace.TestC)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.call("test", [TestA.serialize(arg), TestB.serialize(arg2), Namespace.TestC.serialize(arg3)]);
		}
	}

	public static deserializeArgs(rawArgs: any[]): any[] {
		let argType = TestA.getAlgebraicType();
		let argValue = AlgebraicValue.deserialize(argType, rawArgs[0])
		let arg = TestA.fromValue(argValue);
		let arg2Type = TestB.getAlgebraicType();
		let arg2Value = AlgebraicValue.deserialize(arg2Type, rawArgs[1])
		let arg2 = TestB.fromValue(arg2Value);
		let arg3Type = Namespace.TestC.getAlgebraicType();
		let arg3Value = AlgebraicValue.deserialize(arg3Type, rawArgs[2])
		let arg3 = Namespace.TestC.fromValue(arg3Value);
		return [arg, arg2, arg3];
	}

	public static on(callback: (status: string, identity: string, reducerArgs: any[]) => void)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.on("reducer:Test", callback);
		}
	}
}

__SPACETIMEDB__.reducers.set("Test", TestReducer);
if (__SPACETIMEDB__.spacetimeDBClient) {
	__SPACETIMEDB__.spacetimeDBClient.registerReducer("Test", TestReducer);
}

export default TestReducer
'''
"update_reducer.ts" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

// @ts-ignore
import { __SPACETIMEDB__, AlgebraicType, ProductType, BuiltinType, ProductTypeElement, IDatabaseTable, AlgebraicValue } from "@clockworklabs/spacetimedb-sdk";

export class UpdateReducer
{
	public static call()
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.call("__update__", []);
		}
	}

	public static deserializeArgs(): any[] {
		return [];
	}

	public static on(callback: (status: string, identity: string, reducerArgs: any[]) => void)
	{
		if (__SPACETIMEDB__.spacetimeDBClient) {
			__SPACETIMEDB__.spacetimeDBClient.on("reducer:Update", callback);
		}
	}
}

__SPACETIMEDB__.reducers.set("Update", UpdateReducer);
if (__SPACETIMEDB__.spacetimeDBClient) {
	__SPACETIMEDB__.spacetimeDBClient.registerReducer("Update", UpdateReducer);
}

export default UpdateReducer
'''
